### PR TITLE
Doc k23 si states merged

### PR DESCRIPTION
**Pull request by iafuture**

### CODE REVIEW

File: docs/design/K2_BASIC_CONCEPTS.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+
+## Quick overview
+This is an overview of how data is organized in K2. Details on each part are following
+
+![Data Deployment](../images/Data_Deployment.png)
+
+## Cluster, nodes, nodepool
+A recap from the K2PROCESS.md document:
+A cluster is a set of machines used to deploy data which is managed as a single unit. Wd do not have a notion of cluster in code - it is just how we may refer in documentation to a set of machines which are used to deploy a `Collection`(see below). The data lives in partitions, each deployed to a single CPU Core with some particular fixed memory available to it (4-12GB). We call this a `node` and is what messages in the system are send to. We start these on a given machine as individual threads in a process which we call the `nodepool`. Currently we deploy nodepools one per CPU socket, so a 2-socket machine (each socket with say 20cores) would get 2 nodepools, each running 20 threads for a total of 40 nodes available for deployment on this machine.
```
Comment (jfunston) [inline docs/design/K2_BASIC_CONCEPTS.md:117]: Typos: "Wd"->"We" in second sentence. "messages in the system are send to" -> "messages in the system are sent to".

File: docs/design/K2_BASIC_CONCEPTS.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+## Schema
+A Schema is what we use to describe the data in a data record. In particular, it specifies
+- the names and types for all fields
+- which fields comprise the partitionKey
+- which fields comprise the rangeKey
+
+Schemas have names and are versioned, and is akin to a `table` in the SQL world. All of the data in a particular schema is supposed to be part of a single data set (e.g. `Users`, or `Orders`). Via versioning, the users can perform schema migration, though we do not support changes to the key fields (partition/range).
+
+- Schemas are created and are therefore scoped inside a particular collection.
+- When reading data, the user has to specify from which schema they are reading
```
Comment (jfunston) [inline docs/design/K2_BASIC_CONCEPTS.md:137]: Could link to the schema design docs here if you want.

File: docs/design/K2_BASIC_CONCEPTS.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+## SKVRecord
+In our API, keys and data are not specified explicitly. Instead, the user specifies a set of fields with a schema. The combination of these two is an `SKVRecord`.
+
+The SKVRecord designates a set of fields as the `partitionKey`, and a set of fields as the `rangeKey`. Data will be placed on a given partition based on the `partitionKey` of the record. The `rangeKey` allows records with identical `partitionKey` to be placed adjacent in memory on the same partiiton in order to allow for very efficient bulk retrieval (see reading multiple records below).
```
Comment (jfunston) [inline docs/design/K2_BASIC_CONCEPTS.md:142]: Is it right to say the SKVRecord designates the key fields? To me it is the schema that does that. Typo in last sentence "partiiton"

File: docs/design/K2PROCESS.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+This document describes the architecture for a K2 process.
+
+# General process architecture
+![architecture](../images/K2process.png)
+
+We use Seastar to manage our processes. When our process is started, we create N reactor instances - each instance pinned to a particular core. The number of instances and their core assignment is driven via command line configuration.
+
+The user code is packaged in Applets. An applet is simply any class which has a start and stop methods (see many examples under k2/src/cmd/demo)
```
Comment (jfunston) [inline docs/design/K2PROCESS.md:8]: Maybe a note here that there is one instance of an Applet class per core/reactor.

File: docs/design/K23SI_TXN.md [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+![TWIM_abandoned](../images/TWIM_abandoned.png)
+
+In the TWIM management, if we recover a transaction in an in-progress state, we transition it to the `ForceFinalize` state, where we actively reach out to the TRH to discover the current state of the transaction. Note that it is possible that the transaction is either committed or aborted. Once we find out which one, we apply that transition, moving the txn to `Committed|Aborted` state. At that point we either continue finalizing if we're outside of retention window, or wait for TRH to come-in to finalize things.
+
+# Recovery process
+<mark>TODO</mark> describe handling of failures of TRH, Read participant, Write participant
+
+# Other ideas
+- It may be helpful to allow applications to execute operations in batches so that we can group operations to the same node into single message
+- Consider using separate WAL for intents. Potentially cheaper to GC since we can just maintain a watermarm and drop the tail past the watermark once WIs are finalized. May cause write amplification though
```
Comment (jfunston) [inline docs/design/K23SI_TXN.md:503]: Typo "watermarm"

File: docs/design/K23SI_TXN.md [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+- provide atomic higher-level operations (sinfonia style):
+    - swap
+    - cas
+    - atomic_read_many
+    - acquire_lease
+    - acquire_lease_many
+    - update_if_lease_held
+- We might achieve better throughput under standard benchmark if we consider allowing for a HOLD in cases of conflict resolution(PUSH operation). If we have a Candidate/Pusher which we think will succeed if we knew the outcome of an intent, we can hold onto the candidate operation for short period of time to allow for the intent to commit. For a better implementation, it maybe best to implement a solution which does a transparent hold - a hold that doesn't require special handling at the client (e.g. additional notification and heartbeating). THis could be achieved simply by re-queueing an incoming task once with a delay of potential 999 network round-trip latency (e.g. 10-20usecs).
```
Comment (jfunston) [inline docs/design/K23SI_TXN.md:511]: Typo "THis"

File: docs/design/K23SI_TXN.md [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+- `writeRanges`: As the CL issues write operations during the txn execution, it remembers the partition ranges to which it wrote. This is needed at commit/abort time so that the cluster can contact the partition owners for these ranges, and ask them to convert the WIs they might have on behalf of this transaction. Note that since we support cross-collection transactions, this field has a set of ranges per collection.
+- `trh`: this is the TRH key set by the CL when it first does a write. It is used to determine which partition node currently owns the TR record and is the key of the first write done in the txn
+- `syncFinalize`: this is a txn option which the client can set. It makes sure that a commit/abort isn't acknowledged until all WIs in the cluster have been finalized. This is normally not needed, but it is useful in cases where the client knows that conflicting transactions will be starting next (e.g. doing a large initial data load)
+- `timeToFinalize`: works in conjunction with syncFinalize and specifies an artificial wait to be performed before issuing the finalization call to participants.
```
Comment (jfunston) [inline docs/design/K23SI_TXN.md:313]: Maybe mention that this exists for testing purposes.

File: docs/design/K23SI_TXN.md [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+- When a read is received for a transaction `T`, we remember that the given key was observed at timestamp `T.timestamp`. If we've previously observed a read for the same key, we update the cache to reflect the max such value and the latest transaction to observe it.
+- The implementation of the read key cache is an interval tree. We require an interval tree to also remember key range reads.
+- The cache is consulted at write time to determine if a write should be aborted. The reason we need to do this is that if we try to write an item with timestamp <= lastSVTimeTheKeyWasRead, then we are breaking a promise to whoever read the item - they saw some item version when they read at their snapshot time, and now we're trying to insert a newer version into their snapshot. This write should therefore be aborted.
+- Entries are removed from the cache in an LRU fashion. We maintain a minSVTimestamp watermark for the entire cache, which tells us how old is the oldest entry in the cache. Any write before this timestamp (for any key) is aborted as we assume there may have been a read for it.
```
Comment (jfunston) [inline docs/design/K23SI_TXN.md:164]: Not a big deal, but it is technically not an LRU. We always evict the oldest entry by insertion order, cache hits don't change the eviction order.

**Review by jfunston — CHANGES_REQUESTED** [review]

File: docs/design/K2_BASIC_CONCEPTS.md (hunk 1 above)
Comment (iafuture) [reply docs/design/K2_BASIC_CONCEPTS.md:117]: fixed

File: docs/design/K2_BASIC_CONCEPTS.md (hunk 2 above)
Comment (iafuture) [reply docs/design/K2_BASIC_CONCEPTS.md:137]: linked

File: docs/design/K2_BASIC_CONCEPTS.md (hunk 3 above)
Comment (iafuture) [reply docs/design/K2_BASIC_CONCEPTS.md:142]: You're right. Reworded and fixed typo

File: docs/design/K23SI_TXN.md (hunk 5 above)
Comment (iafuture) [reply docs/design/K23SI_TXN.md:503]: fixed

File: docs/design/K2PROCESS.md (hunk 4 above)
Comment (iafuture) [reply docs/design/K2PROCESS.md:8]: good point. expanded to include this

File: docs/design/K23SI_TXN.md (hunk 6 above)
Comment (iafuture) [reply docs/design/K23SI_TXN.md:511]: fixed

File: docs/design/K23SI_TXN.md (hunk 8 above)
Comment (iafuture) [reply docs/design/K23SI_TXN.md:164]: true. Reworded.

**Review by jfunston — APPROVED** [review]

**Review by jerryhfeng — APPROVED** [review]

**Pull request merged**